yt_dlp/extractor/deezer.py

```diff
@@ -1,92 +1,269 @@
 import json
-
+import re
+from hashlib import md5
 from .common import InfoExtractor
 from ..utils import (
-    ExtractorError,
-    int_or_none,
-    orderedSet,
+    str_to_int,
+    traverse_obj,
 )
 
 
 class DeezerBaseInfoExtractor(InfoExtractor):
+
+    GW_LIGHT_URL = "https://www.deezer.com/ajax/gw-light.php"
+    GET_URL = "https://media.deezer.com/v1/get_url"
+    BLOWFISH_KEY = b"g4el58wc0zvf9na1"
```

Comment:
How is this key obtained?

Reply:
From here : https://github.com/ytdl-org/youtube-dl/pull/13194/files#diff-301769c0cffc70e23f61421e11e68f1c77bf60db2f7547430bd3c4c894609699R40

Reply:
Leaving the key in the repository may not be the best idea. Deezer has sent [DMCAs](https://github.com/github/dmca/blob/master/2021/02/2021-02-10-deezer.md) to many repos containing it in the past. It specifically mentions `"To do so, these repositories are sharing the private encryption keys retrieved maliciously to bypass Deezer's security measures to unlawfully download its music catalogue, directly from Deezer’s servers in total violation of our rights and those of our music licensors (phonographic producers, performing artists, songwriters and composers)."`

It might be a better idea to [dynamically generate the key](https://github.com/yne/dzr/issues/11) or just make users provide it themselves.

Reply:
@ryan5453 I agree, this is a good idea. Note that this key is not directly used to decrypt the song. It is used to compute a new key that will decrypt the song.

Reply:
@ryan5453 I have no idea what [this code](https://github.com/yne/dzr/issues/11#issue-1049202131) returns. I can't run it on my windows. Do you know how to generate this key in python directly ?

Reply:
you could implement it in python with something like this:
```py
    def _get_blowfish_key(self, string):
        hex_str_list = re.findall(r'%5B(0x.{51})%5D', string)
        if len(hex_str_list) != 2:
            return
        list_a, list_b = [list(reversed(hex_str.split('%2C'))) for hex_str in hex_str_list]
        if len(list_a) != 8 or len(list_b) != 8:
            return
        final = []
        for i in range(8):
            final.append(list_a[i])
            final.append(list_b[i])
        return intlist_to_bytes([int(f, base=16) for f in final])

    explore_page = self._download_webpage('https://www.deezer.com/en/channels/explore', None)
    app_web_js = self._download_webpage(
        _search_regex(r'<script src="(https://[\w.-]+/cache/js/app-web\.[a-f0-9]+\.js)',
                      explore_page, 'javascript'), None, note='Downloading webapp JS')
    blowfish_key = self._get_blowfish_key(app_web_js)
```
...which adds an import:
```py
from ..utils import (
    intlist_to_bytes,
    sanitized_Request,
    sanitize_open,
)
```